[MT-1750] Update Facebook SDK (#12)

sebastian-krajna opened the pull request:
- Update Facebook SDK version to 18.0.3
- Update dependencies and configurations across multiple modules including Kotlin 1.9.3 and AGP 8.10.0
- Increase compileSdkVersion and targetSdkVersion from 31 to 33
- Refactor FacebookRemoteCommand class for improved structure and readability
- Enhance unit tests for FacebookRemoteCommand and LogProductItem
- Add client token metadata in AndroidManifest.xml and facebook.json
- Refactor FacebookCommand interface to accept clientToken parameter
- Update Tealium Facebook SDK version to 3.0.0

sebastian-krajna committed:
Update Facebook SDK version to 18.0.3

sebastian-krajna committed:
Update dependencies and configurations across multiple modules

- Upgraded Kotlin version from 1.6.0 to 2.1.0 in build.gradle.
- Updated Android Gradle Plugin from 4.0.2 to 8.10.0.
- Increased compileSdkVersion and targetSdkVersion from 31 to 36 in app and facebook modules.
- Updated various dependencies to their latest versions, including Tealium and AndroidX libraries.
- Refactored MainActivity and TealiumHelper for improved code clarity and consistency.
- Added .vscode/settings.json for Java configuration.
- Adjusted Facebook SDK integration and updated related classes for better structure and readability.
- Updated gradle-wrapper.properties to use Gradle 8.14.

sebastian-krajna committed:
Refactor FacebookRemoteCommand class and enhance unit tests

- Refactored the FacebookRemoteCommand class to improve structure and readability by moving the constructor documentation and adding an additional constructor for auto-initialization.
- Added unit tests for onInvoke method to handle single and multiple commands, ensuring proper invocation of the FacebookCommand methods.
- Implemented tests to verify that logPurchase is not called with NaN amounts and that logProductItem is not called with invalid currency or NaN price amounts.
- Added a test for handling unknown event names in standardEvent method.

sebastian-krajna committed:
Enhance unit tests for FacebookRemoteCommand and LogProductItem

- Added a test to ensure logPurchase is not called with an invalid amount (non-numeric).
- Improved command parsing tests in FacebookRemoteCommand to handle multiple and empty commands.
- Updated LogProductItem tests to safely remove product item parameters using safe calls.
- Removed outdated tests for invalid currency and NaN price amounts in LogProductItem.

sebastian-krajna committed:
Update Facebook SDK integration and configuration

- Upgraded mockk_version from 1.12.8 to 1.13.9 in build.gradle.
- Downgraded compileSdkVersion and targetSdkVersion from 36 to 35 in app and facebook modules for compatibility.
- Added client token metadata in AndroidManifest.xml and facebook.json.
- Updated strings.xml to include facebook_client_token and modified fb_login_protocol_scheme.
- Refactored FacebookCommand interface and related classes to accept clientToken as a parameter for initialization.

sebastian-krajna committed:
Update Tealium Facebook SDK version to 2.0.3 in build.gradle

sebastian-krajna committed:
Remove .vscode/settings.json file as part of project cleanup

jameskeith commented on facebook/build.gradle:
```diff
@@ -31,52 +29,69 @@ android {
-    testImplementation "org.robolectric:robolectric:4.8"
+    testImplementation "org.robolectric:robolectric:4.14.1"
     testImplementation "io.mockk:mockk:$mockk_version"
-    androidTestImplementation 'androidx.test:runner:1.4.0'
-    androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'
-    implementation "androidx.core:core-ktx:1.7.0"
-    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    androidTestImplementation 'androidx.test:runner:1.6.2'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.6.1'
+    implementation "androidx.core:core-ktx:1.16.0"
+    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
```
suspect we can slim these down.
I doubt that we actually depend on:
```
implementation 'androidx.appcompat:appcompat:1.7.0'
implementation "androidx.core:core-ktx:1.16.0"
implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
```

jameskeith commented on facebook/build.gradle:
```diff
@@ -31,52 +29,69 @@ android {
     api "com.facebook.android:facebook-android-sdk:$facebook_version"
-    implementation 'com.tealium:remotecommands:1.0.1'
-    implementation 'androidx.appcompat:appcompat:1.4.1'
+    implementation 'com.tealium:remotecommands:1.0.2'
```
latest versions is actually 1.0.3

jameskeith commented on facebook/src/main/java/com/tealium/remotecommands/facebook/FacebookRemoteCommand.kt:
```diff
@@ -22,48 +11,36 @@ import java.math.BigDecimal
+ * Constructs a RemoteCommand that integrates with the Facebook App Events SDK to allow Facebook API calls to be implemented through Tealium.
+ */
+constructor(
+    application: Application? = null,
```
I realise that this was `nullable` previously, but it looks to be non-functional without it, so can we update it and then tidy up all the unnecessary checks?

jameskeith commented on facebook/src/main/java/com/tealium/remotecommands/facebook/FacebookRemoteCommand.kt:
```diff
@@ -152,6 +127,10 @@ class FacebookRemoteCommand : RemoteCommand {
                         if (payload.optString(Initialize.APPLICATION_ID).isNotBlank()) payload.optString(
                             Initialize.APPLICATION_ID
                         ) else null
+                    val clientToken =
+                        if (payload.optString(Initialize.CLIENT_TOKEN).isNotBlank()) payload.optString(
+                            Initialize.CLIENT_TOKEN
+                        ) else null
```
I think we can make this a little cleaner and the lines around it
```kotlin
val clientToken = payload.optString(Initialize.CLIENT_TOKEN).ifBlank { null }
```

jameskeith commented on facebook/src/main/java/com/tealium/remotecommands/facebook/FacebookRemoteCommand.kt:
```diff
@@ -188,7 +168,7 @@ class FacebookRemoteCommand : RemoteCommand {
                 Commands.SET_USER_ID -> {
                     val userId = payload.optString(User.USER_ID)
-                    userId?.let {
+                    userId.let {
```
for the non-nullable variables, we can just remove the `let` and get rid of an extra level of nesting

sebastian-krajna committed:
Update project dependencies and configurations

- Downgraded Kotlin version from 2.1.0 to 1.9.23 in build.gradle.
- Updated Facebook SDK version range to [7,19) and upgraded Tealium Facebook SDK to 3.0.0.
- Downgraded compileSdkVersion and targetSdkVersion from 35 to 33 in app and facebook modules for compatibility.
- Changed Java compatibility from version 21 to 11 in app and facebook modules.
- Added .vscode/settings.json for Java build configuration.
- Updated gradle-wrapper.properties to use Gradle 7.6.3.

sebastian-krajna committed:
Remove .vscode/settings.json file as part of project cleanup

sebastian-krajna committed:
Refactor FacebookRemoteCommand constructor to simplify initialization

- Made the application parameter private and removed unnecessary initialization code.
- Streamlined the constructor by eliminating optional parameters for Facebook application ID, client token, and debug flag.

jameskeith commented on build.gradle:
```diff
@@ -1,18 +1,18 @@
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:4.0.2'
+        classpath 'com.android.tools.build:gradle:7.4.2'
```
Feels like we've rowed a little too far back on this one - I think you were right to move this to 8.x before, but since targeting API 35 is what we're after (for now) to meet play store requirements for a while, then AGP v8.6.0 is probably best idea since it's the min for API 35 and maintains support for Android Studio as far back as [Koala Feature Drop](https://developer.android.com/build/releases/gradle-plugin)

jameskeith commented on facebook/build.gradle:
```diff
@@ -1,22 +1,20 @@
-
-        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
-        consumerProguardFiles 'consumer-rules.pro'
+        targetSdkVersion 33
```
linked to the AGP comment, I think you were right with 35 for now. 👍

jameskeith commented on facebook/src/main/java/com/tealium/remotecommands/facebook/FacebookRemoteCommand.kt:
```diff
@@ -330,30 +281,22 @@ class FacebookRemoteCommand : RemoteCommand {
 
-    fun logPurchase(purchase: JSONObject) {
+    private fun logPurchase(purchase: JSONObject) {
         val amount = purchase.optDouble(Purchase.PURCHASE_AMOUNT) as? Double
         amount?.let {
             if (it.isNaN()) {
```
Whilst we're here, can we tidy this up as well? `optDouble` always returns a non-null `double` so this can be simplified:
```
val amount = purchase.optDouble(Purchase.PURCHASE_AMOUNT)
if (amount.isNaN()) {
    return
}
```
I believe there's another unnecessary cast like this further down this file as well.

jameskeith reviewed (commented):
some really nice changes here, great work - couple of thoughts from me and I think we're good to go

sebastian-krajna commented on facebook/src/main/java/com/tealium/remotecommands/facebook/FacebookRemoteCommand.kt:
```diff
@@ -22,48 +11,23 @@ import java.math.BigDecimal
-    /**
-     * Constructs a RemoteCommand that integrates with the Facebook App Events SDK to allow Facebook API calls to be implemented through Tealium.
-     */
-    @JvmOverloads
-    constructor(
-        application: Application? = null,
-        commandId: String = DEFAULT_COMMAND_ID,
-        description: String = DEFAULT_COMMAND_DESCRIPTION,
-        facebookApplicationId: String? = null,
-        debugEnabled: Boolean? = null
-    ) : super(commandId, description, BuildConfig.TEALIUM_FACEBOOK_VERSION) {
-        application?.let { app ->
-            this.application = app
-            facebookApplicationId?.let {
-                facebookInstance = FacebookInstance(app, it, debugEnabled)
-            }
-        }
-    }
```
The autoInit constructor didn't actually do anything useful. The existing constructor with facebookApplicationId already provides early initialization when needed. It properly creates the FacebookInstance object with the app ID when provided. If someone needs early SDK setup, they can just use the main constructor with their Facebook app ID.

jameskeith replied:
I know we discussed the `autoInit` one being removed as it wasn't doing anything. But in the latest branch there is only a single `constructor` and it doesn't accept an App Id.

I've highlighted this specific removal because it looks to be the one that _was_ providing the auto-init behaviour.

sebastian-krajna replied:
You're right that I removed both the autoInit constructor (which wasn't functional) and the constructor with facebookApplicationId parameter.
After reviewing the code more thoroughly, I noticed that Facebook initialization is primarily meant to happen through the Commands.INITIALIZE command path. The INITIALIZE command properly sets up facebookInstance with all required parameters.
Having two ways to initialize (constructor and command) creates redundancy and potential confusion. Users might initialize in the constructor but then later try to re-initialize through commands, creating inconsistent behavior.
Looking at the example usage in TealiumHelper, it uses the simple constructor and relies on commands for initialization. This seems to be the intended pattern.
That said, if users definitely need auto-initialization at construction time, we could restore that constructor. Would you prefer I add it back while keeping the command-based initialization as well?

sebastian-krajna committed:
Upgrade project dependencies and configurations

- Updated Android Gradle Plugin from 7.4.2 to 8.6.0.
- Increased compileSdkVersion and targetSdkVersion from 33 to 35 in app and facebook modules.
- Changed Java compatibility from version 11 to 17 in app and facebook modules.
- Updated gradle-wrapper.properties to use Gradle 8.7.

sebastian-krajna committed:
Fix BigDecimal initialization in FacebookRemoteCommand to ensure proper conversion from amount string

sebastian-krajna committed:
Update Java compatibility in build.gradle files for app and facebook modules from version 17 to 11

sebastian-krajna committed:
Refactor FacebookRemoteCommand constructor to enforce non-null application parameter

- Changed the application parameter from nullable to non-nullable.
- Simplified initialization logic by removing unnecessary null checks and directly initializing FacebookInstance if application ID and client token are provided.

jameskeith commented on facebook/src/test/java/com/tealium/remotecommands/facebook/FacebookRemoteCommandTest.kt:
```diff
@@ -20,15 +13,19 @@ import org.junit.Test
 import org.robolectric.RobolectricTestRunner
 import java.util.*
+import android.app.Application
+import org.robolectric.RuntimeEnvironment
```
this appears to be unused

jameskeith reviewed (approved):
two really small updates to go, and then lets get this released

sebastian-krajna committed:
Reorder parameters in FacebookRemoteCommand constructor for clarity

- Moved commandId and description parameters to the end of the constructor for better readability and consistency with default values.
- Removed unused import in FacebookRemoteCommandTest to clean up the code.
